### PR TITLE
Clean up long functions in virt-controller migration.go

### PR DESCRIPTION
**Pull request by prnaraya**
**What this PR does / why we need it**: This PR cleans up two very long functions, `sync` and `updateStatus` in virt-controller `migration.go`, extracting code snippets in placing them into separate functions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: The [Boy Scout Rule](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html) was considered and applied if needed
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to the [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/watch/migration.go [hunk 1]
```diff
@@ -1342,6 +1340,44 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 	return nil
 }
 
+func (c *MigrationController) setupTargetPodCreation(vmi *virtv1.VirtualMachineInstance) error {
```
Comment (vasiliy-ul) [inline pkg/virt-controller/watch/migration.go:1343]: The function name is too generic and does not reflect the real purpose. I would name it more explicitly, e.g.:

```suggestion
func (c *MigrationController) setupVMIRuntimeUser(vmi *virtv1.VirtualMachineInstance) error {
```

File: pkg/virt-controller/watch/migration.go [hunk 2]
```diff
@@ -429,6 +429,15 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
 
+	migrationPhaseArgs := &processMigrationPhaseArgs{
+		Pod:                 pod,
+		AttachmentPod:       attachmentPod,
+		PodExists:           podExists,
+		AttachmentPodExists: attachmentPodExists,
+		ConditionManager:    conditionManager,
+		VmiConditionManager: vmiConditionManager,
+	}
+
```
Comment (vasiliy-ul) [inline pkg/virt-controller/watch/migration.go:440]: Can we create this struct right before calling the `processMigrationPhase` function? IMHO, it is not clear why this struct is needed here.

File: pkg/virt-controller/watch/migration.go [hunk 3]
```diff
@@ -604,6 +534,102 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
+type processMigrationPhaseArgs struct {
+	Pod                 *k8sv1.Pod
+	AttachmentPod       *k8sv1.Pod
+	PodExists           bool
+	AttachmentPodExists bool
+	ConditionManager    *controller.VirtualMachineInstanceMigrationConditionManager
+	VmiConditionManager *controller.VirtualMachineInstanceConditionManager
+}
+
+func (c *MigrationController) processMigrationPhase(args *processMigrationPhaseArgs, migration, migrationCopy *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance, syncError error) error {
```
Comment (vasiliy-ul) [inline pkg/virt-controller/watch/migration.go:546]: Why do we pack only some of the args into a struct? Why not all?

**Review by vasiliy-ul — COMMENTED** [review]
/retest-required

**Comment by vasiliy-ul** [comment]
There are quite some failures related to migration: pull-kubevirt-e2e-k8s-1.27-sig-compute-migrations

File: pkg/virt-controller/watch/migration.go [hunk 4]
```diff
@@ -501,88 +510,9 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
-				}
-			}
-
-			if vmi.Status.MigrationState.Completed &&
-				!vmiConditionManager.HasCondition(vmi, virtv1.VirtualMachineInstanceVCPUChange) {
-				migrationCopy.Status.Phase = virtv1.MigrationSucceeded
-				c.recorder.Eventf(migration, k8sv1.EventTypeNormal, SuccessfulMigrationReason, "Source node reported migration succeeded")
-				log.Log.Object(migration).Infof("VMI reported migration succeeded.")
-			}
+		err := c.processMigrationPhase(migrationPhaseArgs, migration, migrationCopy, vmi, syncError)
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go:513]: Yeah, personally I was not really sure whether in this case it is better to have a long function signature with each parameter listed out or to have a struct to contain the parameters. I know that [sonar cloud](https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&id=kubevirt_kubevirt) flags functions with more than 7 parameters so that is why I chose to use the struct based on what sonar cloud was flagging as code smell.

File: pkg/virt-controller/watch/migration.go [hunk 5]
```diff
@@ -604,6 +526,92 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
+		}
+	case virtv1.MigrationPending:
+		if podExists {
+			if controller.VMIHasHotplugVolumes(vmi) {
+				if attachmentPodExists {
+					migrationCopy.Status.Phase = virtv1.MigrationScheduling
+				}
+			} else {
+				migrationCopy.Status.Phase = virtv1.MigrationScheduling
+			}
```
Comment (alicefr) [inline pkg/virt-controller/watch/migration.go:554]: Can we simplify this? You are always setting the same value for the Phase

File: pkg/virt-controller/watch/migration.go [hunk 6]
```diff
@@ -604,6 +526,92 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
+		}
+		if isPodReady(pod) {
+			if controller.VMIHasHotplugVolumes(vmi) {
+				if attachmentPodExists && isPodReady(attachmentPod) {
+					log.Log.Object(migration).Infof("Attachment pod %s for vmi %s/%s is ready", attachmentPod.Name, vmi.Namespace, vmi.Name)
+					migrationCopy.Status.Phase = virtv1.MigrationScheduled
+				}
+			} else {
+				migrationCopy.Status.Phase = virtv1.MigrationScheduled
+			}
```
Comment (alicefr) [inline pkg/virt-controller/watch/migration.go:575]: Here again, the Phase is always the same, could you please reduce to a single statement?

Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go:575]: I was able to reduce the code from your first comment but I am not sure exactly how to reduce this snippet of code while preserving the logging of the attachment pod being ready.

**Comment by prnaraya** [comment]
@vasiliy-ul @alicefr Would you be able to take another look?

**Review by vasiliy-ul — COMMENTED** [review]
@prnaraya, overall the code looks good. I've added a couple of suggestions inline about returning nil on error and wrapping the function arguments.

Also commented regarding simplification of the conditions. My personal preference is to keep it in the original form. But as mentioned, I do not insist on reverting the change. If you and @alicefr both agree that this way is better, then please ignore those comments.

File: pkg/virt-controller/watch/migration.go [hunk 7]
```diff
@@ -603,6 +524,89 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
+			// in progress for this VMI.
+			migrationCopy.Status.Phase = virtv1.MigrationFailed
+			c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "VMI is not eligible for migration because another migration job is in progress.")
+			log.Log.Object(migration).Error("Migration object ont eligible for migration because another job is in progress")
+		}
+	case virtv1.MigrationPending:
+		if podExists {
+			if !controller.VMIHasHotplugVolumes(vmi) || attachmentPodExists {
+				migrationCopy.Status.Phase = virtv1.MigrationScheduling
+			}
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go:548]: I agree that the original logic flows much nicer here than the changed version which is hard to follow (and therefore doesn't improve readability), so I am inclined to revert both this if statement and the other one to their original forms. My only hangup is with the duplicated lines when setting the migration phase, which I think were the motivation for changing the if statements here in the first place.

[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go:548]: I agree that the original logic flows much nicer here than the changed version which is hard to follow (and therefore doesn't improve readability), so I am inclined to revert both this if statement and the other one to their original forms. My only hangup is with the duplicated lines when setting the migration phase (like `migrationCopy.Status.Phase = virtv1.MigrationScheduling`), which I think were the motivation for changing the if statements here in the first place. Do you have any opinions about this?

File: pkg/virt-controller/watch/migration.go [hunk 8]
```diff
@@ -603,6 +524,89 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 	return nil
 }
 
+func (c *MigrationController) processMigrationPhase(migration, migrationCopy *virtv1.VirtualMachineInstanceMigration, pod, attachmentPod *k8sv1.Pod, podExists, attachmentPodExists bool, conditionManager *controller.VirtualMachineInstanceMigrationConditionManager, vmiConditionManager *controller.VirtualMachineInstanceConditionManager, vmi *virtv1.VirtualMachineInstance, syncError error) (*virtv1.VirtualMachineInstanceMigration, error) {
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/migration.go [hunk 9]
```diff
@@ -603,6 +524,89 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
...
+		_, exists := pod.Annotations[virtv1.MigrationTargetReadyTimestamp]
+		if !exists && vmi.Status.MigrationState.TargetNodeDomainReadyTimestamp != nil {
+			key := patch.EscapeJSONPointer(virtv1.MigrationTargetReadyTimestamp)
+			patchOps := fmt.Sprintf(`[{ "op": "add", "path": "/metadata/annotations/%s", "value": "%s" }]`,
+				key,
+				vmi.Status.MigrationState.TargetNodeDomainReadyTimestamp.String())
+
+			_, err := c.clientset.CoreV1().Pods(pod.Namespace).Patch(context.Background(), pod.Name, types.JSONPatchType, []byte(patchOps), v1.PatchOptions{})
+			if err != nil {
+				return migrationCopy, err
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/migration.go [hunk 10]
```diff
@@ -603,6 +524,89 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 	return nil
 }
 
+func (c *MigrationController) processMigrationPhase(migration, migrationCopy *virtv1.VirtualMachineInstanceMigration, pod, attachmentPod *k8sv1.Pod, podExists, attachmentPodExists bool, conditionManager *controller.VirtualMachineInstanceMigrationConditionManager, vmiConditionManager *controller.VirtualMachineInstanceConditionManager, vmi *virtv1.VirtualMachineInstance, syncError error) (*virtv1.VirtualMachineInstanceMigration, error) {
+	switch migration.Status.Phase {
+	case virtv1.MigrationPhaseUnset:
+		canMigrate, err := c.canMigrateVMI(migration, vmi)
+		if err != nil {
+			return migrationCopy, err
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

**Comment by vasiliy-ul** [comment]
@prnaraya, thank you for addressing the comments. From my perspective, the PR looks good. 

/approve
/retest-required

File: pkg/virt-controller/watch/migration.go [hunk 11]
```diff
@@ -500,7 +500,7 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 		log.Log.Object(migration).Errorf("target attachment pod %s/%s shutdown during migration", attachmentPod.Namespace, attachmentPod.Name)
 	} else {
 		var err error
-		migration, migrationCopy, err = c.processMigrationPhase(migration, migrationCopy, pod, attachmentPod, podExists, attachmentPodExists, conditionManager, vmiConditionManager, vmi, syncError)
+		migrationCopy, err = c.processMigrationPhase(migration, migrationCopy, pod, attachmentPod, podExists, attachmentPodExists, conditionManager, vmiConditionManager, vmi, syncError)
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:503]: Can you elaborate on what is happening in `Attempting to fix sig-compute-migration tests. ` ?

File: pkg/virt-controller/watch/migration.go [hunk 12]
```diff
@@ -603,6 +524,89 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 	return nil
 }
 
+func (c *MigrationController) processMigrationPhase(migration, migrationCopy *virtv1.VirtualMachineInstanceMigration, pod, attachmentPod *k8sv1.Pod, podExists, attachmentPodExists bool, conditionManager *controller.VirtualMachineInstanceMigrationConditionManager, vmiConditionManager *controller.VirtualMachineInstanceConditionManager, vmi *virtv1.VirtualMachineInstance, syncError error) (updatedMigration, updatedMigrationCopy *virtv1.VirtualMachineInstanceMigration, err error) {
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/migration.go (hunk 11 above)
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/migration.go [hunk 13]
```diff
@@ -524,12 +524,20 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 	return nil
 }
 
-func (c *MigrationController) processMigrationPhase(migration, migrationCopy *virtv1.VirtualMachineInstanceMigration, pod, attachmentPod *k8sv1.Pod, podExists, attachmentPodExists bool, conditionManager *controller.VirtualMachineInstanceMigrationConditionManager, vmiConditionManager *controller.VirtualMachineInstanceConditionManager, vmi *virtv1.VirtualMachineInstance, syncError error) (*virtv1.VirtualMachineInstanceMigration, error) {
```
[the remark this answers is not in the document]
Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/migration.go [hunk 14]
```diff
@@ -1260,7 +1274,7 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 			}
 
 			// patch VMI annotations and set RuntimeUser in preparation for target pod creation
-			err = c.setupTargetPodCreation(vmi)
+			err = c.setupVMIRuntimeUser(vmi)
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go]: This should have been done in the first commit.

File: pkg/virt-controller/watch/migration.go [hunk 15]
```diff
@@ -1274,9 +1274,13 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 			}
 
 			// patch VMI annotations and set RuntimeUser in preparation for target pod creation
-			err = c.setupVMIRuntimeUser(vmi)
-			if err != nil {
-				return err
+			patches := c.setupVMIRuntimeUser(vmi)
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:1277]: This should be part of the first commit

**Review by xpivarc — COMMENTED** [review]
Almost there

File: pkg/virt-controller/watch/migration.go [hunk 16]
```diff
@@ -1259,35 +1259,10 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
...
-					if _, ok := vmi.Annotations[virtv1.DeprecatedNonRootVMIAnnotation]; ok {
-						patches = append(patches, fmt.Sprintf(`{ "op": "remove", "path": "/metadata/annotations/%s"}`, patch.EscapeJSONPointer(virtv1.DeprecatedNonRootVMIAnnotation)))
-					}
-				}
-			}
+			// patch VMI annotations and set RuntimeUser in preparation for target pod creation
+			patches := c.setupVMIRuntimeUser(vmi)
 			if len(patches) != 0 {
-				vmi, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(patches), &v1.PatchOptions{})
+				_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(patches), &v1.PatchOptions{})
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:1278]: Why do we make this change?

Comment (prnaraya) [reply pkg/virt-controller/watch/migration.go]: Sorry, I must have missed that while I was rebasing. I've reverted this change.

**Review by xpivarc — APPROVED** [review]
